### PR TITLE
Use terra application layout

### LINKED ISSUE #56: Uplift terra-dev-site to consume terra-application-layout
# Issue Description
Terra-dev-site should consume terra-application-layout (https://github.com/cerner/terra-framework/issues/77). This should eliminate a bunch of custom components in terra-dev-site.

## Issue Type

- [ ] New Feature
- [x] Enhancement
- [ ] Bug
- [ ] Other

### PR DESCRIPTION
**Pull request by mjhenkes**
### Summary

Resolves #56 

#### Changes
* Consume terra-application-layout, remove custom layout
* Consume RoutingMenu from terra-application-layout
* A GitHub link extension can be added through config.

This is passive except for terra-framework, where it breaks tests by fixing translations.

@cerner/terra

[CONTRIBUTORS.md]: https://github.com/cerner/terra-dev-site/blob/master/CONTRIBUTORS.md
[License]: https://github.com/cerner/terra-dev-site/blob/master/License.md

These PRS demonstrate this components consumption.
https://github.com/cerner/terra-framework/pull/127
https://github.com/cerner/terra-clinical/pull/314
https://github.com/cerner/terra-core/pull/1374
https://github.com/cerner/terra-ui/pull/78

### CODE REVIEW

File: src/components/GitHubLinkExtension.jsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */
+import React from 'react';
+import PropTypes from 'prop-types';
+import Image from 'terra-image';
+
+const propTypes = {
+  /**
+   * The children list items passed to the component.
```
Comment (emilyrohrbough) [inline src/components/GitHubLinkExtension.jsx]: Nit: description needs updated

File: src/components/Extensions.scss [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+  .container-compact {
+    background-color: #d3d3d3;
+    display: flex;
+    padding: 10px;
+    width: 100%;
+  }
+
+  .item {
+    margin-right: 5px;
```
Comment (emilyrohrbough) [inline src/components/Extensions.scss]: update to rem?

File: src/app/components/RawRoute.jsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { matchPath } from 'react-router-dom';
+
+const RawRoute = (routingConfig, location) => {
+  const routes = Object.keys(routingConfig.content).sort();
+  const nonRawPath = location.pathname.substring(4);
```
[the remark this answers is not in the document]
Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: I want to trim '/raw' from the string. Is there a cleaner way?

File: src/app/App.jsx [hunk 4]
```diff
@@ -92,73 +89,77 @@ class App extends React.Component {
...
     return (
-      <ThemeProvider id="site" themeName={this.props.themes[this.state.theme]} isGlobalTheme>
-        <Base className="base" locale={this.state.locale}>
-          <NavigationLayout
-            header={applicationHeader}
-            menuText="Menu"
-            indexPath={this.props.navigation && this.props.navigation.index}
-            config={this.props.routeConfig}
-          />
+      <ThemeProvider id="site" themeName={appConfig.themes[theme]} isGlobalTheme>
```
Comment (emilyrohrbough) [inline src/app/App.jsx]: `themeName={appConfig.themes[theme]} ` I think we lost the passing of appConfig/themes in the props and we are now always using the default theme as the global theme. Consumer theme defines their own theme as the default

File: index.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import GitHubLinkExtension from './src/components/GitHubLinkExtension';
```
Comment (emilyrohrbough) [inline index.js]: Should we also export the Components and MenuComponent ? Better naming might be useful

File: index.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+import Extensions from './src/components/Extensions';
+
+export { GitHubLinkExtension };
```
Comment (emilyrohrbough) [inline index.js]: nit: should be able to write `export { GitHubLinkExtension , Extensions };`

File: src/Index.jsx [hunk 7]
```diff
@@ -31,25 +33,34 @@ class Site extends React.Component {
   render() {
     const siteConfig = Object.assign({}, defaultSiteConfig, this.state.siteConfig);
     siteConfig.appConfig = Object.assign({}, defaultSiteConfig.appConfig, siteConfig.appConfig);
-    const { appConfig, navConfig, componentConfig } = siteConfig;
+    const { appConfig, componentConfig } = siteConfig;
 
     const { routeConfig, navigation } = routeConfiguration(siteConfig, componentConfig);
 
     const routes = Object.freeze(routeConfig);
+
```
Comment (emilyrohrbough) [inline src/Index.jsx]: Since this is the entry point for the dev-site, to me it would make sense to move the creation of appLogo and nameConfig to the `App` component

File: src/Index.jsx [hunk 8]
```diff
@@ -31,25 +33,34 @@ class Site extends React.Component {
...
+      title: appConfig.title,
+    };
+
     return (
       <Router>
         <App
-          routeConfig={routes}
-          navigation={navigation}
-          rootPath={navConfig.rootPath}
-          themes={appConfig.themes}
```
Comment (emilyrohrbough) [inline src/Index.jsx]: Think we will want to keep themes

File: index.js (hunk 5 above)
Comment (mjhenkes) [reply index.js]: Exporting the template makes sense. I'm making changes to not export extensions and the github extension in favor of building it with app config. I don't want to export components because no one should consume it outside of terra-dev-site.

File: index.js (hunk 6 above)
Comment (mjhenkes) [reply index.js]: One would think, but that wasn't working.

File: src/app/configureApp.jsx [hunk 9]
```diff
@@ -1,71 +1,70 @@
...
+  let menuComponent = RoutingMenu;
+  if (link.menuComponent) {
+    menuComponent = link.menuComponent;
+  }
 
-  subNavConfig.forEach((subConfig) => {
-    generatedConfig[subConfig.path] = subConfig;
-  });
+  // Manipulate the link config to build out the first level menu item.
+  const { generatedConfig } = buildMenuConfig({
```
Comment (emilyrohrbough) [inline src/app/configureApp.jsx]: This is a little hard to read, maybe break out this first object to pass in?

File: src/app/configureApp.jsx [hunk 10]
```diff
@@ -1,71 +1,70 @@
...
-
-      // Do not create a submenu for the component if the component has one site page with no additional sub-nav.
-      if (exampleType !== 'tests' && examples.length === 1 && !examples[0][`${exampleType}`]) {
-        return undefined;
+const buildMenuConfig = (component, menuComponent, exampleType, pathRoot = '') => {
+  let generatedConfig = {};
+  const path = pathRoot + component.path;
+  const examples = component[exampleType];
+
+  if (examples) {
```
Comment (emilyrohrbough) [inline src/app/configureApp.jsx]: Does this create links for config where a page has a component but not examples?

-this git diff is hard to follow

File: src/app/configureApp.jsx [hunk 11]
```diff
@@ -1,71 +1,70 @@
...
-    )
-  )
-);
-
 const buildComponent = (Component, configuredProps) => (
   {
     default: {
-      componentClass: injectConfig(configuredProps)(Component),
+      componentClass: Component,
+      props: configuredProps,
```
Comment (emilyrohrbough) [inline src/app/configureApp.jsx]: Eventually we will probably want to open these config to allow props to be passed with the component that we combine with ours

Comment (mjhenkes) [reply src/app/configureApp.jsx]: Yeah I was thinking about adding that, but we don't have a case for it yet.

File: src/components/GitHubLinkExtension.jsx [hunk 12]
```diff
@@ -0,0 +1,35 @@
+/* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */
```
Comment (emilyrohrbough) [inline src/components/GitHubLinkExtension.jsx]: We should remove this linter comment. You were exporting the GitHubLinkExtension above in index.js so all our dependencies should be defined

File: src/app/App.jsx [hunk 13]
```diff
@@ -92,73 +89,77 @@ class App extends React.Component {
...
+        <Base className="base" locale={locale}>
+          <Switch>
+            <Route
+              path="/raw"
+              render={() => RawRoute(routingConfig, location)}
```
Comment (emilyrohrbough) [inline src/app/App.jsx]: With how this is set up, you are creating a raw route for each page, not just the test pages: https://terra-framework-deploye-pr-127.herokuapp.com/static/#/raw/site/home

File: src/app/App.jsx [hunk 14]
```diff
@@ -92,73 +89,77 @@ class App extends React.Component {
...
       this.setState({ theme: nextProps.defaultTheme });
     }
 
-    if (nextProps.defaultDir !== undefined && nextProps.defaultDir !== this.props.defaultDir) {
+    if (App.propExistsAndChanged(nextProps.defaultDir, this.props.defaultDir)) {
       document.getElementsByTagName('html')[0].setAttribute('dir', nextProps.defaultDir);
       this.setState({ dir: nextProps.defaultDir });
     }
+
+    this.utilityConfig = this.setupUtilityConfig(nextProps.utilityConfig);
```
Comment (emilyrohrbough) [inline src/app/App.jsx]: Do we need to setup the utility config each time the app receive props?

File: src/app/App.jsx (hunk 13 above)
Comment (mjhenkes) [reply src/app/App.jsx]: This isn't creating any routes, just allowing any route defined to be accessed without app framework. This is intentional.

File: src/app/App.jsx (hunk 14 above)
Comment (mjhenkes) [reply src/app/App.jsx]: Yes, this applies the onChange and metadata instance functions to the static config provided externally.

File: src/app/components/RawRoute.jsx [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+  const route = routes.find(routeToMatch => matchPath(nonRawPath, routeToMatch));
+
+  if (route) {
+    const routeData = routingConfig.content[route].component.default;
+    const ComponentClass = routeData.componentClass;
+    const componentProps = Object.assign({}, routeData.props, { pathRoot: `/raw${route}` });
+    return <ComponentClass {...componentProps} />;
+  }
+  return 404;
```
Comment (emilyrohrbough) [inline src/app/components/RawRoute.jsx]: We need a terra crash page :D

Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: Agreed, but returning the text of 404 was cheaper for now. :)

Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: @neilpfeiffer wanna make us a sweet 404 page?

File: src/app/components/RawRoute.jsx [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
```
Comment (emilyrohrbough) [inline src/app/components/RawRoute.jsx]: I'm thinking it might make more sense to either only render Raw Test Routes or possibly use a key to indicate raw routes are desired. at this point, only tests seems fine

Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: I'd prefer to just allow everything. I'd like remove special logic around the test routes.

Comment (emilyrohrbough) [reply src/app/components/RawRoute.jsx]: Did you choose to render each route here for the 404 result instead of the site redirect?

File: src/app/components/RawRoute.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (emilyrohrbough) [reply src/app/components/RawRoute.jsx]: Ah, haven't seen it done that way.  The ways I've seen is using replace or splice but neither of the three seem better or worse

File: src/app/components/RawRoute.jsx (hunk 16 above)
Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: I honestly think we shouldn't redirect on error. If this was a test page, i'd rather see a 404 in a screenshot than the home page. It'd be more obvious what the problem was.

Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: I'd like to change the component to also render a 404 if the route isn't found.

File: src/app/ConfigureUtilities.jsx [hunk 17]
```diff
@@ -0,0 +1,112 @@
+import React from 'react';
+import IconSettings from 'terra-icon/lib/icon/IconSettings';
+
+function generateItemConfig(defaultItem, items, key) {
```
Comment (emilyrohrbough) [inline src/app/ConfigureUtilities.jsx]: Typically we've been writing utility functions with the arrow syntax as
```
const generateItemConfig = (defaultItem, items, key) => {
    ...
}
```

File: src/app/ConfigureUtilities.jsx [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+    Object.keys(parentMenuItem.childKeys).forEach((item) => {
+      // eslint-disable-next-line no-param-reassign
+      parentMenuItem.childKeys[item].isSelected = item === selectedItem;
+    });
+  }
+}
+
+class ConfigureUtilties {
+
+  static generateInitialUtiltiesConfig(appConfig) {
```
Comment (emilyrohrbough) [inline src/app/ConfigureUtilities.jsx]: you should be able to move these static functions to use the arrow syntax above and then to improve readability on what functions are exported, we could update to be like:
```
const ConfigureUtilties = {
    generateInitialUtiltiesConfig,
    convertChildkeysToArray,
   ... etc
};

export default ConfigureUtilties;
```

Comment (mjhenkes) [reply src/app/ConfigureUtilities.jsx]: I prefer the class syntax.

File: src/app/configureApp.jsx (hunk 11 above)
Comment (emilyrohrbough) [reply src/app/configureApp.jsx]: In working on the docs for dev-site I noticed it with the menu, but we've removed the menu here so I think its fine to leave until there is need!

File: src/app/ConfigureUtilities.jsx (hunk 17 above)
Comment (mjhenkes) [reply src/app/ConfigureUtilities.jsx]: Why is that preferred?

Comment (emilyrohrbough) [reply src/app/ConfigureUtilities.jsx]: es6 syntax? I'm not sure, its just the pattern we've followed in other component's utils ([Table](https://github.com/cerner/terra-core/blob/master/packages/terra-table/src/SelectableUtils.js), [hookshot](https://github.com/cerner/terra-core/blob/master/packages/terra-hookshot/src/_HookshotUtils.js), [app-layout](https://github.com/cerner/terra-framework/blob/master/packages/terra-application-layout/src/utils/utilityHelpers.jsx), ect)

File: src/app/components/RawRoute.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjhenkes) [reply src/app/components/RawRoute.jsx]: I'll pass in the string to remove from the location, so this looks less weird.

File: index.js (hunk 5 above)
Comment (mjhenkes) [reply index.js]: I'm going to leave exporting the template up to @Blackop778 with his work on upgrading the template.

File: src/Index.jsx (hunk 7 above)
Comment (mjhenkes) [reply src/Index.jsx]: I'm trying to separate any static config outside of the `App` component. Long term I want to generate any static config prior to running webpack and creating the app. The app name and logo are static.

File: src/app/configureApp.jsx (hunk 10 above)
Comment (mjhenkes) [reply src/app/configureApp.jsx]: This both creates and adds sub menus to the flat list of menu routes and adds items to the current menu.

File: navigation.config.js [hunk 19]
```diff
@@ -15,6 +15,15 @@ const navConfig = {
         menuText: 'Getting Started',
         subMenuText: 'Info',
       },
+    }, {
+      path: '/dev-site/getting-started/thing',
```
Comment (ryanthemanuel) [inline navigation.config.js]: Thing?

File: src/app/components/GitHubLinkExtension.jsx [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+const defaultProps = {
+  href: undefined,
+};
+
+const GithubLinkExtension = (props) => {
+  const { href } = props;
+  return (<a href={href}>
+    <Image
+      variant="rounded"
+      src={'https://github.com/cerner/terra-dev-site/raw/master/src/app/components/github-logo.png'}
```
Comment (ryanthemanuel) [inline src/app/components/GitHubLinkExtension.jsx]: Anything special copyright wise associated with this image?

Comment (mjhenkes) [reply src/app/components/GitHubLinkExtension.jsx]: We should be good. 
>Use the Octocat or GitHub logo to link to GitHub

https://github.com/logos

File: navigation.config.js (hunk 19 above)
Comment (mjhenkes) [reply navigation.config.js]: whoops, some debug code slipped through.

File: src/app/components/RawRoute.jsx (hunk 15 above)
Comment (neilpfeiffer) [reply src/app/components/RawRoute.jsx]: Sounds good - I'll see what I can do.

File: index.js [hunk 21]
```diff
@@ -0,0 +1 @@
+
```
Comment (emilyrohrbough) [inline index.js]: should we remove for now?

File: src/app/configureApp.jsx [hunk 22]
```diff
@@ -1,71 +1,75 @@
-import React from 'react';
```
Comment (emilyrohrbough) [inline src/app/configureApp.jsx]: should this be .js now?

File: site.config.js [hunk 23]
```diff
@@ -22,6 +22,9 @@ const siteConfig = {
     title: 'Terra Dev Site',
     locales: ['en'],
     bidirectional: false,
+    extensions: {
```
Comment (emilyrohrbough) [inline site.config.js]: Right now, the default site config is the best documentation we have on the site options. We should add the `extensions` key and a quick blrub on its usage and assign a default.
https://github.com/cerner/terra-dev-site/blob/master/src/config/site.config.js#L4

File: src/app/configureApp.jsx (hunk 22 above)
Comment (mjhenkes) [reply src/app/configureApp.jsx]: yes 👍

File: index.js (hunk 21 above)
Comment (mjhenkes) [reply index.js]: I'd like to keep it since I link to it from the package.json

File: src/app/App.jsx [hunk 24]
```diff
@@ -68,20 +61,26 @@ const propTypes = {
...
 const defaultProps = {
-  appTitle: appConfig.title,
-  appLogoSrc: appConfig.logoSrc,
-  hideBidiUtility: !appConfig.bidirectional,
-  defaultDir: appConfig.defaultDirection,
-  defaultTheme: appConfig.defaultTheme,
-  themes: appConfig.themes,
-  defaultLocale: appConfig.defaultLocale,
-  locales: appConfig.locales,
+  nameConfig: undefined,
```
Comment (tbiethman) [inline src/app/App.jsx]: Do we need these defaults? Undefined is the implicit default value.

Comment (mjhenkes) [reply src/app/App.jsx]: linter complains without it

Comment (tbiethman) [reply src/app/App.jsx]: 😞

File: site.config.js (hunk 23 above)
Comment (mjhenkes) [reply site.config.js]: I added extensions to the default site config with a blurb about usage. I have it commented out because there's no 'right' github url default.

File: src/app/App.jsx (hunk 24 above)
Comment (mjhenkes) [reply src/app/App.jsx]: This rule is disabled in core etc, but not here.
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/require-default-props.md

**Comment by bjankord** [comment]
delete
